### PR TITLE
Reduce LLVM IR codegen logic complexities

### LINKED ISSUE #68: Organize compiler src folder
**Description:**
Move the files instead compiler folder in two at least two sub folders. One for BIR parser side logic and one for CodeGen side logic

**Suggested Labels:**
Type/Improvement

**Suggested Assignees:**
ruvi-d

### LINKED ISSUE #69: Fixing compiler warnings
**Description:**
Current master is throws a few warnings even with `-Wall` OFF. They should be addressed so that we can enable `-Werror` for release builds.

**Suggested Labels:**
Type/Improvement

**Suggested Assignees:**
@shubhamnarlawar @ramakotareddy

### LINKED ISSUE #114: Reduce translate function complexities
**Description:**
Currently the codegen class's translate functions have a lot of nested if/else checks to test the invariants of the objects and not all the logic path's are being handled. This is leading to compiler and static analysis warnings. We can improve this by removing the default class constructors and removing most setters. This would force the class objects invariants to be well defined; allowing us to skip the invariant testing in in the translate functions. This will improve performance, fix the compiler warnings and also improve the code maintainability.

**Suggested Labels:**
Priority/High, Type/Improvement, Points/2

**Suggested Assignees:**
Ruvi-d

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**
#69

### LINKED ISSUE #140: Fix source code formatting
**Description:**
The CPP and H files don't follow our guidelines of using 4 spaces for indentation and 120 character line lengths.
Need to fix the formatting on all files and add a format config for this codebase. 

**Suggested Labels:**
Type/Improvement

**Suggested Assignees:**
Ruvi-d

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### PR DESCRIPTION
**Pull request by ruvi-d**
## Purpose
Resolves #114, Resolves #69, Resolves #68, Resolves #140 

**Change summary:**
* Improve speed of source compile time
  * 5s vs 20s on my laptop
* Improved compiler warnings
  * Enable extra warnings for debug builds
  * Fix warnings generated for debug and release builds
  * Enable treat warnings as errors for release builds
* Refactor and restructure codegen classes
  * Major changes were done to the said class hierarchy
  * The target was to guarantee class invariance so that the codegen logic requires less checks
  * Cast instruction logic was not refactored much due to the type string and union type support PRs that are awaiting to e merged
* Format all CPP and H files according to code convention
  * Added `.clang-format` file to repo that can be used by clang-format and VS code IDE to format this codebase
  * Add license header to all cpp and h files
* Not addressed in the PR
  * Refactoring BIR parser classes 
  * Resolving issues with passing Strings by value
  * Memory leaks

## Approach
* The file names with llvm prefix were renamed to Bal to more accurately reflect the fact that It is modeling the BIR structure in memory
* These files were also moved to a separate `codegen` sub folder inside the `compiler` folder
* All these classes were moved inside a new namespace called `nballerina` 
  * I've taken the liberty with the class names, feel free to raise any concerns with them
* Removed most of the function definitions that were in the header
* Removed most of the dead code paths and functions that were never used
* The catch all `BIR.h` file has been broken into smaller header files
  * Extensive use of forward declaration was used to break cyclic dependencies 
* The VarDecl, Operand and other variable related classes saw major overhauls to their class hierarchy 
  * Cast instruction and ConstLoad Instructions were left relatively unchanged
* Used `default`, `override` and `final` keywords were applicable

**Note**: I have tried to stick to the standard [C++ coding guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines) as much as possible. Please do review and catch any I've missed.

### CODE REVIEW

**Comment by Kishanthan** [comment]
Lets call for a review session on this soon. @ruvi-d

**Comment by ruvi-d** [comment]
> Lets call for a review session on this soon. @ruvi-d

Scheduled for Monday (08/03/2021) 10:30AM IST

File: compiler/BIRReader.cpp [hunk 1]
```diff
@@ -29,962 +50,824 @@ ReadMapStoreInsn ReadMapStoreInsn::readMapStoreInsn;
...
-  }
-  default:
-    fprintf(stderr, "%s:%d Invalid Insn Kind for Reader.\n", __FILE__,
-            __LINE__);
-    break;
-  }
-  if (nonTerminatorInsn)
-    basicBlock->addNonTermInsn(nonTerminatorInsn);
+void BIRReader::readInsn(BasicBlock *basicBlock) {
+
```
Comment (Kishanthan) [inline compiler/BIRReader.cpp:468]: Unwanted new line?

File: compiler/BIRReader.cpp [hunk 2]
```diff
@@ -29,962 +50,824 @@ ReadMapStoreInsn ReadMapStoreInsn::readMapStoreInsn;
...
-      default:
-        fprintf(stderr, "%s:%d Invalid Insn Kind for Instruction Patching.\n",
-                __FILE__, __LINE__);
-        break;
-      }
+void BIRReader::patchInsn(vector<BasicBlock *> basicBlocks) {
+    for (auto const &basicBlock : basicBlocks) {
+        Function *curFunc = basicBlock->getFunction();
+        TerminatorInsn *terminator = basicBlock->getTerminatorInsn();
+        if (terminator && terminator->getPatchStatus()) {
```
Comment (Kishanthan) [inline compiler/BIRReader.cpp:598]: Inverse the if here?

File: compiler/BIRReader.cpp [hunk 3]
```diff
@@ -29,962 +50,824 @@ ReadMapStoreInsn ReadMapStoreInsn::readMapStoreInsn;
...
-  }
-  case TYPE_TAG_NIL:
-    constantloadInsn->setTypeTagNil(typeTag);
-  default:
-    break;
-  }
-  constantloadInsn->setLhsOperand(lhsOperand);
-  return constantloadInsn;
+ConstantLoadInsn *ReadConstLoadInsn::readNonTerminatorInsn(BasicBlock *currentBB) {
+
```
Comment (Kishanthan) [inline compiler/BIRReader.cpp:290]: Unwanted new line?

**Review by Kishanthan — APPROVED** [review]

**Pull request merged**